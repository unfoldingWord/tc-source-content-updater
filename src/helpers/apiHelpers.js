import {delay} from './utils';
import {SORT, STAGE, SUBJECT} from '../index';

const request = require('request');
const DOOR43_CATALOG = `Door43-Catalog`;

/**
 * does http request and returns the response data parsed from JSON
 * @param {string} url
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function makeJsonRequestDetailed(url, retries=5) {
  let result_;
  for (let i = 1; i <= retries; i++) {
    result_ = null;
    try {
      result_ = await new Promise((resolve, reject) => {
        request(url, function(error, response, body) {
          if (error)
            reject(error);
          else if (response.statusCode === 200) {
            let result = body;
            try {
              result = JSON.parse(body);
            } catch (e) {
              reject(e);
            }
            resolve({result, response, body});
          } else {
            reject(`fetch error ${response.statusCode}`);
          }
        });
      });
    } catch (e) {
      if (i >= retries) {
        console.warn('makeJsonRequestDetailed() - error getting manifest data', e);
        throw e;
      }
      result_ = null;
      await delay(500);
      console.log(`makeJsonRequestDetailed() - retry ${i+1} getting manifest data, last error`, e);
    }

    if (result_) {
      break;
    }
  }
  return result_;
}

/**
 * searching subjects
 * @param {array} subjects
 * @param {string} owner
 * @param {number} retries
 * @return {Promise<*>}
 */
async function searchSubjects(subjects, owner, retries=3) {
  const subjectParam = encodeURI(subjects.join(','));
  let fetchUrl = `https://git.door43.org/api/catalog/v3/search?subject=${subjectParam}`;
  if (owner) {
    fetchUrl += fetchUrl + `&owner=${owner}`;
  }
  const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
  return result;
}

/**
 * get published catalog using catalog next v3
 * @return {Promise<*[]>}
 */
export async function getOldCatalogReleases() {
  const released = [];
  const owner = null; // get all owners
  const subjectList = ['Bible', 'Aligned Bible', 'Greek New Testament', 'Hebrew Old Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];
  // const subjectList = ['Bible', 'Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];

  try {
    // for (const subject of subjectList) {
    const result = await searchSubjects(subjectList, owner, 5);
    let repos = 0;
    const languages = result && result.languages || [];
    for (const language of languages) {
        const languageId = language.identifier;
        const resources = language.resources || [];
        for (const resource of resources) {
          resource.languageId = languageId;
          resource.resourceId = resource.identifier;
          resource.foundInCatalog = 'OLD';
          resource.full_name = resource.full_name || `${resource.owner}/${resource.repo}`;
          resource.checking_level = resource.checking && resource.checking.checking_level;
          released.push(resource);
          repos++;
        }
      }
      console.log(`has ${repos} items`);
    // }
    console.log(`released catalog has ${released.length} items`);
  } catch (e) {
    console.error('getCatalog() - error getting catalog', e);
    return [];
  }

  return released;
}

/**
 * get published catalog
 * @return {Promise<*[]>}
 */
export async function getCatalog() {
  const catalogReleases = await getOldCatalogReleases();
  const searchParams = {
    subject: SUBJECT.ALL_TC_RESOURCES,
    stage: STAGE.PROD,
  };
  console.log(`found ${catalogReleases.length} items in old catalog`);
  const newCatalogReleases = await searchCatalogNext(searchParams);
  console.log(`found ${newCatalogReleases.length} items in new catalog`);
  // merge catalogs together - catalog new takes precident
  for (const item of newCatalogReleases) {
    const index = catalogReleases.findIndex(oldItem => (item.full_name === oldItem.full_name));
    if (index >= 0) {
      catalogReleases[index] = item; // overwrite item in old catalog
      catalogReleases[index].foundInCatalog = 'NEW+OLD';
    } else {
      catalogReleases.push(item); // add unique item
    }
  }
  console.log(`now ${catalogReleases.length} items in merged catalog`);

  return catalogReleases;
}

/**
 * add parameter to url string
 * @param {*} value
 * @param {object} parameters
 * @param {string} tag
 * @return {*}
 */
function addUrlParameter(value, parameters, tag) {
  if (value) {
    if (parameters) { // see if we need separator
      parameters += '&';
    }
    parameters += `${tag}=${encodeURIComponent(value)}`;
  }
  return parameters;
}

/**
 * filter for supported repos
 * @param {array} resources
 * @return {*[]}
 */
function getCompatibleResourceList(resources) {
  const supported = [];
  for (const item of resources || []) {
    // add fields for backward compatibility
    const languageId = item.language;
    let [, resourceId] = (item.name || '').split(`${languageId}_`);
    resourceId = resourceId || item.name; // if language was not in name, then use name as resource ID
    item.resourceId = resourceId;
    item.languageId = languageId;
    item.checking_level = item.repo && item.repo.checking_level;
    item.foundInCatalog = 'NEW';
    item.modified = item.modified || item.released;

    if (item.zipball_url) {
      item.downloadUrl = item.zipball_url;
    }
    // check for version. if there is one, it will save having to fetch it from DCS later.
    if (item.release) { // if released
      const tagName = item.release.tag_name;
      if (tagName && (tagName[0] === 'v')) {
        item.version = tagName.substr(1);
      }
    } else {
      const branchOrTagName = item.branch_or_tag_name;
      if (branchOrTagName && (branchOrTagName[0] === 'v')) {
        item.version = branchOrTagName.substr(1);
      }
    }
    if (item.subject) {
      item.subject = item.subject.replaceAll(' ', '_');
    }
    // add supported resources to returned list
    if (item.downloadUrl && item.subject && item.name && item.full_name) {
      supported.push(item);
    }
  }
  return supported;
}

/**
 * get published catalog
 * @param {Object} searchParams - details below
 * @param {String} searchParams.owner - if undefined then all are searched
 * @param {String} searchParams.languageId - if undefined then all are searched
 * @param {String} searchParams.subject - one or more separated by comma.  If undefined then all are searched.
 *          Example `Bible,Aligned Bible,Greek New Testament,Hebrew Old Testament,Translation Words,TSV Translation Notes,Translation Academy'
 * @param {Number} searchParams.limit - maximum results to return, default 100
 * @param {Boolean} searchParams.partialMatch - if true will do case insensitive substring matching, default is false
 * @param {String} searchParams.stage - specifies which release stage to be returned out of these stages:
 *                    "prod" - return only the production releases (default)
 *                    "preprod" - return the pre-production release if it exists instead of the production release
 *                    "draft" - return the draft release if it exists instead of pre-production or production release
 *                   "latest" -return the default branch (e.g. master) if it is a valid RC instead of the "prod", "preprod" or "draft".  (default)
 * @param {Number} searchParams.checkingLevel - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {Number} searchParams.sort - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {number} retries - number of times to retry calling search API, default 3
 * @return {Promise<*[]|null>}
 */
export async function searchCatalogNext(searchParams, retries=3) {
  let result_ = null;
  const {
    owner,
    languageId,
    subject,
    limit = 100,
    stage = STAGE.LATEST,
    checkingLevel,
    partialMatch,
    sort = SORT.REPO_NAME,
  } = searchParams;

  try {
    let fetchUrl = `https://git.door43.org/api/catalog/v5/search`;
    let parameters = '';
    parameters = addUrlParameter(owner, parameters, 'owner');
    parameters = addUrlParameter(languageId, parameters, 'lang');
    parameters = addUrlParameter(subject, parameters, 'subject');
    parameters = addUrlParameter(limit, parameters, 'limit');
    parameters = addUrlParameter(stage, parameters, 'stage');
    parameters = addUrlParameter(checkingLevel, parameters, 'checkingLevel');
    parameters = addUrlParameter(partialMatch, parameters, 'partialMatch');
    parameters = addUrlParameter(sort, parameters, 'sort');
    if (parameters) {
      fetchUrl += '?' + parameters;
    }
    console.log(`Searching: ${fetchUrl}`);
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    result_ = result && result.data;
  } catch (e) {
    console.warn('searchCatalogNext() - error calling search API', e);
    return null;
  }

  const supported = getCompatibleResourceList(result_);
  return supported;
}

/**
 * does Catalog next API query to get manifest data
 * @param {string} owner
 * @param {string} repo
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function downloadManifestData(owner, repo, retries=5) {
  const fetchUrl = `https://git.door43.org/api/catalog/v5/entry/${owner}/${repo}/master/metadata`;
  try {
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    return result;
  } catch (e) {
    console.warn('getManifestData() - error getting manifest data', e);
    throw e;
  }
}
