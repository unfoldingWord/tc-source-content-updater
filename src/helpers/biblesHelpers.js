/* eslint-disable camelcase */
import fs from 'fs-extra';
import path from 'path-extra';

/**
 * generate manifest.json
 * @param {Object} oldManifest - old manifest data
 * @param {String} RESOURCE_OUTPUT_PATH - folder to store manifest.json
 * @param {Object} resource
 * @return {Object} new manifest data
 */
export function generateBibleManifest(oldManifest, RESOURCE_OUTPUT_PATH, resource) {
  const newManifest = {};
  newManifest.dublin_core = oldManifest.dublin_core; // preserve original manifest data
  newManifest.checking = oldManifest.checking;
  newManifest.projects = oldManifest.projects;
  newManifest.original_manifest = oldManifest;

  // copy some data for more convenient access
  newManifest.language_id = oldManifest.dublin_core.language.identifier;
  newManifest.language_name = oldManifest.dublin_core.language.title;
  newManifest.direction = oldManifest.dublin_core.language.direction;
  newManifest.subject = oldManifest.dublin_core.subject;
  newManifest.resource_id = oldManifest.dublin_core.identifier;
  newManifest.resource_title = oldManifest.dublin_core.title;
  const remoteModifiedTime = resource && resource.remoteModifiedTime || (new Date()).toJSON();
  newManifest.remoteModifiedTime = remoteModifiedTime;
  const oldMainfestIdentifier = oldManifest.dublin_core.identifier
                                      .toLowerCase();
  const identifiers = ['ugnt', 'ubh'];
  newManifest.description = identifiers.includes(oldMainfestIdentifier) ?
    'Original Language' : 'Gateway Language';

  const savePath = path.join(RESOURCE_OUTPUT_PATH, 'manifest.json');
  fs.writeFileSync(savePath, JSON.stringify(newManifest, null, 2));
  return newManifest;
}
