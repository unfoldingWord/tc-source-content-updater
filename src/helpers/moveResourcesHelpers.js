import fs from 'fs-extra';
import path from 'path';

/**
 * @description transfer an entire resource from source to target directory
 * @param {string} resourceSourcePath - current position of resource
 * @param {string} resourceTargetPath - folder where resources are moved
 * @return {Promise} Move directory Promise
 */
export function moveResources(resourceSourcePath, resourceTargetPath) {
  return new Promise((resolve, reject) => {
    if (resourceSourcePath && resourceSourcePath.length &&
            resourceTargetPath && resourceTargetPath.length &&
            fs.pathExistsSync(resourceSourcePath)) {
      if (fs.pathExistsSync(resourceTargetPath)) {
        fs.removeSync(resourceTargetPath);
      }
      fs.ensureDirSync(path.dirname(resourceTargetPath));
      fs.copySync(resourceSourcePath, resourceTargetPath);
      fs.removeSync(resourceSourcePath);
      resolve(resourceTargetPath);
    } else {
      reject('Invalid parameters to moveResources()');
    }
  });
}

