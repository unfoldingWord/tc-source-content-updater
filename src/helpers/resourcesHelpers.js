/* eslint-disable no-console,max-len,camelcase */
import fs from 'fs-extra';
import path from 'path-extra';
import yaml from 'yamljs';
import {isObject} from 'util';
// helpers
import * as zipFileHelpers from './zipFileHelpers';
import * as twArticleHelpers from './translationHelps/twArticleHelpers';
import * as tnArticleHelpers from './translationHelps/tnArticleHelpers';
import * as taArticleHelpers from './translationHelps/taArticleHelpers';
import * as twGroupDataHelpers from './translationHelps/twGroupDataHelpers';
import * as packageParseHelpers from './packageParseHelpers';
// constants
import * as errors from '../resources/errors';
import * as Bible from '../resources/bible';

const translationHelps = {
  ta: 'translationAcademy',
  tn: 'translationNotes',
  tw: 'translationWords',
  tq: 'translationQuestions',
};

/**
 * @description - Gets the version from the manifest
 * @param {String} resourcePath - folder for manifest.json or yaml
 * @return {String} version
 */
export function getVersionFromManifest(resourcePath) {
  const manifest = getResourceManifest(resourcePath);
  if (!manifest || !manifest.dublin_core || !manifest.dublin_core.version) {
    return null;
  }
  return manifest.dublin_core.version;
}

/**
 * @description Helper function to get manifest file from the resources folder. First
 * it will try manifest.json, then manifest.yaml.
 * @param {String} resourcePath - path to a resource folder which contains the manifest file in its root.
 * @return {Object} manifest
 */
export function getResourceManifest(resourcePath) {
  const manifest = getResourceManifestFromJson(resourcePath);
  if (!manifest) {
    return getResourceManifestFromYaml(resourcePath);
  }
  return manifest;
}

/**
 * @description - Turns a manifest.json file into an object and returns it, null if doesn't exist
 * @param {String} resourcePath - folder for manifest.json
 * @return {Object} manifest
 */
export function getResourceManifestFromJson(resourcePath) {
  const fileName = 'manifest.json';
  const manifestPath = path.join(resourcePath, fileName);
  let manifest = null;
  if (fs.existsSync(manifestPath)) {
    manifest = fs.readJsonSync(manifestPath);
  }
  return manifest;
}

/**
 * @description - Turns a manifest.yaml file into an object and returns it, null if doesn't exist
 * @param {String} resourcePath - folder for manifest.yaml
 * @return {Object} manifest
 */
export function getResourceManifestFromYaml(resourcePath) {
  const fileName = 'manifest.yaml';
  const manifestPath = path.join(resourcePath, fileName);
  let manifest = null;
  if (fs.existsSync(manifestPath)) {
    const yamlManifest = fs.readFileSync(manifestPath, 'utf8').replace(/^\uFEFF/, '');
    manifest = yaml.parse(yamlManifest);
  }
  return manifest;
}

/**
 * Returns an array of versions found in the path that start with [vV]\d
 * @param {String} resourcePath - base path to search for versions
 * @return {Array} - array of versions, e.g. ['v1', 'v10', 'v1.1']
 */
export function getVersionsInPath(resourcePath) {
  if (!resourcePath || !fs.pathExistsSync(resourcePath)) {
    return null;
  }
  const isVersionDirectory = (name) => {
    const fullPath = path.join(resourcePath, name);
    return fs.lstatSync(fullPath).isDirectory() && name.match(/^v\d/i);
  };
  return sortVersions(fs.readdirSync(resourcePath).filter(isVersionDirectory));
}

/**
 * Returns a sorted an array of versions so that numeric parts are properly ordered (e.g. v10a < v100)
 * @param {Array} versions - array of versions unsorted: ['v05.5.2', 'v5.5.1', 'V6.21.0', 'v4.22.0', 'v6.1.0', 'v6.1a.0', 'v5.1.0', 'V4.5.0']
 * @return {Array} - array of versions sorted:  ["V4.5.0", "v4.22.0", "v5.1.0", "v5.5.1", "v05.5.2", "v6.1.0", "v6.1a.0", "V6.21.0"]
 */
export function sortVersions(versions) {
  // Don't sort if null, empty or not an array
  if (!versions || !Array.isArray(versions)) {
    return versions;
  }
  // Only sort of all items are strings
  for (let i = 0; i < versions.length; ++i) {
    if (typeof versions[i] !== 'string') {
      return versions;
    }
  }
  versions.sort((a, b) => String(a).localeCompare(b, undefined, {numeric: true}));
  return versions;
}

/**
 * Return the full path to the highest version folder in resource path
 * @param {String} resourcePath - base path to search for versions
 * @return {String} - path to highest version
 */
export function getLatestVersionInPath(resourcePath) {
  const versions = sortVersions(getVersionsInPath(resourcePath));
  if (versions && versions.length) {
    return path.join(resourcePath, versions[versions.length - 1]);
  }
  return null; // return illegal path
}

/**
 * @description Unzips a resource's zip file to an imports directory for processing
 * @param {Object} resource Resource object containing resourceId and languageId
 * @param {String} zipFilePath Path to the zip file
 * @param {string} resourcesPath Path to the resources directory
 * @return {String} Path to the resource's import directory
 */
export const unzipResource = async (resource, zipFilePath, resourcesPath) => {
  const importsPath = path.join(resourcesPath, 'imports');
  fs.ensureDirSync(importsPath);
  const importPath = zipFilePath.split('.').slice(0, -1).join('.');
  await zipFileHelpers.extractZipFile(zipFilePath, importPath);
  return importPath;
};

/**
 * Gets the single subdirectory of an extracted zip file path
 * @param {String} extractedFilesPath Extracted files path
 * @return {String} The subdir in the extracted path
 */
export function getSubdirOfUnzippedResource(extractedFilesPath) {
  const subdirs = fs.readdirSync(extractedFilesPath);
  if (subdirs.length === 1 && fs.lstatSync(path.join(extractedFilesPath, subdirs[0])).isDirectory()) {
    return path.join(extractedFilesPath, subdirs[0]);
  }
  return extractedFilesPath;
}

/**
 * @description Processes a resource in the imports directory as needed
 * @param {Object} resource Resource object
 * @param {String} sourcePath Path to the source dictory of the resource
 * @param {String} resourcesPath Path to user resources folder
 * @return {String} Path to the directory of the processed files
 * @param {Array} downloadErrors - parsed list of download errors with details such as if the download completed (vs. parsing error), error, and url
 */
export async function processResource(resource, sourcePath, resourcesPath, downloadErrors) {
  try {
    if (!resource || !isObject(resource) || !resource.languageId || !resource.resourceId) {
      throw Error(formatError(resource, errors.RESOURCE_NOT_GIVEN));
    }
    if (!sourcePath) {
      throw Error(formatError(resource, errors.SOURCE_PATH_NOT_GIVEN));
    }
    if (!fs.pathExistsSync(sourcePath)) {
      throw Error(formatError(resource, errors.SOURCE_PATH_NOT_EXIST));
    }

    const processedFilesPath = sourcePath + '_processed';
    fs.ensureDirSync(processedFilesPath);

    switch (resource.subject) {
      case 'Translation_Words':
        twArticleHelpers.processTranslationWords(resource, sourcePath, processedFilesPath);
        break;
      case 'TSV_Translation_Notes':
        await tnArticleHelpers.processTranslationNotes(resource, sourcePath, processedFilesPath, resourcesPath, downloadErrors);
        break;
      case 'Translation_Academy':
        taArticleHelpers.processTranslationAcademy(resource, sourcePath, processedFilesPath);
        break;
      case 'Bible':
      case 'Aligned_Bible':
      case 'Greek_New_Testament':
      case 'Hebrew_Old_Testament':
        packageParseHelpers.parseBiblePackage(resource, sourcePath, processedFilesPath);
        break;
      default:
        fs.copySync(sourcePath, processedFilesPath);
    }

    const manifest = getResourceManifest(sourcePath);

    if (!getResourceManifest(processedFilesPath) && manifest) {
      manifest.catalog_modified_time = resource.remoteModifiedTime;
      fs.outputJsonSync(path.join(processedFilesPath, 'manifest.json'), manifest, {spaces: 2});
    }

    return processedFilesPath;
  } catch (error) {
    console.error(error);
    throw Error(appendError(errors.UNABLE_TO_DOWNLOAD_RESOURCES, error));
  }
}

/**
 * @description Gets the actual path to a resource based on the resource object
 * @param {Object} resource The resource object
 * @param {String} resourcesPath The path to the resources directory
 * @return {String} The resource path
 */
export function getActualResourcePath(resource, resourcesPath) {
  const languageId = resource.languageId;
  let resourceName = resource.resourceId;
  let type = 'bibles';
  if (translationHelps[resourceName]) {
    resourceName = translationHelps[resourceName];
    type = 'translationHelps';
  }
  const actualResourcePath = path.join(resourcesPath, languageId, type, resourceName, 'v' + resource.version);
  fs.ensureDirSync(actualResourcePath);
  return actualResourcePath;
}

/**
 * @description Downloads the resources that need to be updated for a given language using the DCS API
 * @param {Object.<{
 *             languageId: String,
 *             resourceId: String,
 *             localModifiedTime: String,
 *             remoteModifiedTime: String,
 *             downloadUrl: String,
 *             version: String,
 *             subject: String,
 *             catalogEntry: {langResource, bookResource, format}
 *           }>} resource - resource to download
 * @param {String} sourcePath Path to the Bible directory
 * @return {String} Path to the processed tw Group Data files
 */
export function makeTwGroupDataResource(resource, sourcePath) {
  if (!resource) {
    throw Error(formatError(resource, errors.RESOURCE_NOT_GIVEN));
  }
  if (!fs.pathExistsSync(sourcePath)) {
    throw Error(formatError(resource, errors.SOURCE_PATH_NOT_EXIST));
  }
  if ((resource.languageId === Bible.NT_ORIG_LANG && resource.resourceId === Bible.NT_ORIG_LANG_BIBLE) ||
      (resource.languageId === Bible.OT_ORIG_LANG && resource.resourceId === Bible.OT_ORIG_LANG_BIBLE)) {
    const twGroupDataPath = path.join(sourcePath + '_tw_group_data_' + resource.languageId + '_v' + resource.version);
    const result = twGroupDataHelpers.generateTwGroupDataFromAlignedBible(resource, sourcePath, twGroupDataPath);
    if (result) {
return twGroupDataPath
;
}
  }
}

/**
 * Removes all version directories except the latest
 * @param {String} resourcePath Path to the resource directory that has subdirs of versions
 * @param {array} versionsToNotDelete List of versions not to be deleted.
 * @return {Boolean} True if versions were deleted, false if nothing was touched
 */
export function removeAllButLatestVersion(resourcePath, versionsToNotDelete = []) {
  // Remove the previoius verison(s)
  const versionDirs = getVersionsInPath(resourcePath);
  if (versionDirs && versionDirs.length > 1) {
    const lastVersion = versionDirs[versionDirs.length - 1];
    versionDirs.forEach((versionDir) => {
      if (versionDir !== lastVersion && !versionsToNotDelete.includes(versionDir)) {
        const versionPath = path.join(resourcePath, versionDir);
        console.log(`removeAllButLatestVersion() - removing old resource: ${versionPath}`);
        fs.removeSync(versionPath);
      }
    });
    return true;
  }
  return false;
}

/**
 * @description Formats an error for all resources to have a standard format
 * @param {Object} resource Resource object
 * @param {String} errMessage Error message
 * @return {String} The formatted error message
 */
export function formatError(resource, errMessage) {
  if (!resource || !isObject(resource) || !resource.languageId || !resource.resourceId) {
    resource = {
      languageId: 'unknown',
      resourceId: 'unknown',
    };
  }
  return resource.languageId + '_' + resource.resourceId + ': ' + errMessage;
}

/**
 *  converts error to string
 * @param {Error|String} error - error to append
 * @return {string} concatenated message
 */
export function getErrorMessage(error) {
  return ((error && error.message) || error || 'UNDEFINED');
}

/**
 * appends error message to string
 * @param {string} str - string to use as prefix
 * @param {Error|String} err - error to append
 * @return {string} concatenated message
 */
export function appendError(str, err) {
  return str + ': ' + getErrorMessage(err);
}

/**
 * Determines if the rootpath plus a filename is a directory.
 * @param {string} rootPath
 * @param {string} filename
 * @return {bool} - Whether the path is a directory or not.
 */
export const isDirectory = (rootPath, filename) => {
  const fullPath = path.join(rootPath, filename);
  return fs.lstatSync(fullPath).isDirectory();
};
