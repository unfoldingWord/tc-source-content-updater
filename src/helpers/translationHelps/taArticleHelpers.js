/* eslint-disable curly */
import { isObject } from 'util';
import fs from 'fs-extra';
import path from 'path-extra';
// heleprs
import * as resourcesHelpers from '../resourcesHelpers';
// constants
import * as errors from '../../resources/errors';

/**
 * @description Processes the extracted files for translationAcademy to create a single file for each
 * article
 * @param {Object} resource - Resource object
 * @param {String} sourcePath - Path to the extracted files that came from the zip file in the catalog
 * @param {String} outputPath - Path to place the processed files WITHOUT version in the path
 * @return {Boolean} true if success
 */
export function processTranslationAcademy(resource, sourcePath, outputPath) {
  if (!resource || !isObject(resource) || !resource.languageId || !resource.resourceId)
    throw Error(resourcesHelpers.formatError(resource, errors.RESOURCE_NOT_GIVEN));

  if (!sourcePath)
    throw Error(resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_GIVEN));

  if (!fs.pathExistsSync(sourcePath))
    throw Error(resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_EXIST));

  if (!outputPath)
    throw Error(resourcesHelpers.formatError(resource, errors.OUTPUT_PATH_NOT_GIVEN));

  if (fs.pathExistsSync(outputPath))
    fs.removeSync(outputPath);
  const manifest = resourcesHelpers.getResourceManifest(sourcePath);

  manifest.projects.forEach((project) => {
    const folderPath = path.join(sourcePath, project.path);
    const isDirectory = (item) => fs.lstatSync(path.join(folderPath, item)).isDirectory();
    const articleDirs = fs.readdirSync(folderPath).filter(isDirectory);

    articleDirs.forEach((articleDir) => {
      const titlePath = path.join(folderPath, articleDir, 'title.md');

      if (!fs.existsSync(titlePath)) {
        console.warn(`processTranslationAcademy() - title missing: ${titlePath}`);
        return;
      }

      let content = '# ' + fs.readFileSync(titlePath, 'utf8') + ' #\n';
      const articlePath = path.join(folderPath, articleDir, '01.md');

      if (!fs.existsSync(articlePath)) {
        console.warn(`processTranslationAcademy() - file missing: ${articlePath}`);
        return;
      }
      content += fs.readFileSync(articlePath, 'utf8');
      const destinationPath = path.join(
        outputPath,
        project.path,
        articleDir + '.md'
      );
      fs.outputFileSync(destinationPath, content);
    });
  });
  return true;
}
