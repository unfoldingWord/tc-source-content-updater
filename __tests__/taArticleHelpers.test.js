import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
import { convertTsv9to7 } from 'tsv-groupdata-parser';
// helpers
import * as taArticleHelpers from '../src/helpers/translationHelps/taArticleHelpers';
import * as resourcesHelpers from '../src/helpers/resourcesHelpers';
// constants
import * as errors from '../src/resources/errors';

describe('Tests for taArticleHelpers', function() {
  const resource = {
    languageId: 'en',
    resourceId: 'ta',
    version: '9'
  };

  beforeEach(() => {
    fs.__resetMockFS();
  });

  // __tests__/fixtures/translationHelps/tnExtractedFromCDN/en_tn/en_tn_57-TIT.tsv
  // convertTsv9to7(tsv_)

  it('test convertTsv9to7', () => {
    const actualFilePath = path.join(__dirname, 'fixtures/translationHelps/tnExtractedFromCDN/en_tn/en_tn_57-TIT.tsv');
    const tsv = fs.__actual.readFileSync(actualFilePath, 'utf8');

    // import { convertTsv9to7 } from 'tsv-groupdata-parser';
    const result = convertTsv9to7(tsv);

    // then
    expect(result.tsv).toBeTruthy();
    expect(result.errors).toBeFalsy();
  });

  it('Test taArticleHelpers.processTranslationAcademy() for en', () => {
    // given
    const actualExtractedPath = path.join(__dirname, 'fixtures/translationHelps/taExtractedFromCDN');
    const mockedExtractedPath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta');
    fs.__loadDirIntoMockFs(actualExtractedPath, mockedExtractedPath);
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta_processed');
    fs.ensureDirSync(outputPath);
    const expectedProjectList = ['checking', 'translate'];
    const expectedCheckingArticleListLength = 6;
    const expectedTranslateArticleListLength = 5;

    // when
    const result = taArticleHelpers.processTranslationAcademy(resource, path.join(mockedExtractedPath, resource.languageId + '_' + resource.resourceId), outputPath);
    const projectList = fs.readdirSync(outputPath);
    const checkingArticleList = fs.readdirSync(path.join(outputPath, 'checking'));
    const translateArticleList = fs.readdirSync(path.join(outputPath, 'translate'));
    const whatisFile = path.join(outputPath, 'translate', 'translate-whatis.md');
    const whatisArticle = fs.readFileSync(whatisFile, 'utf8');

    // then
    expect(result).toBeTruthy();
    expect(fs.existsSync(whatisFile)).toBeTruthy();
    expect(projectList.sort()).toEqual(expectedProjectList);
    expect(checkingArticleList.length).toEqual(expectedCheckingArticleListLength);
    expect(checkingArticleList).toMatchSnapshot();
    expect(translateArticleList.length).toEqual(expectedTranslateArticleListLength);
    expect(translateArticleList).toMatchSnapshot();
    expect(whatisArticle).toMatchSnapshot();
  });

  it('Test taArticlesHelpers.processTranslationAcademy() for invalid source path', () => {
    // given
    const sourcePath = '/bad/dir';
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta_processed');
    const expectedError = resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_EXIST);

    // when
    expect(() => taArticleHelpers.processTranslationAcademy(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test taArticlesHelpers.processTranslationAcademy() for no source path given', () => {
    // given
    const sourcePath = null;
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta_processed');
    const expectedError = errors.SOURCE_PATH_NOT_GIVEN;

    // when
    expect(() => taArticleHelpers.processTranslationAcademy(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test taArticlesHelpers.processTranslationAcademy() for no resource given', () => {
    // given
    const sourcePath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta');
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta_processed');
    const resource = null;
    fs.ensureDirSync(sourcePath);
    const expectedError = errors.RESOURCE_NOT_GIVEN;

    // when
    expect(() => taArticleHelpers.processTranslationAcademy(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test taArticlesHelpers.processTranslationAcademy() for invalid output path', () => {
    // given
    const sourcePath = path.join(ospath.home(), 'translationCore/resources/imports/en_ta');
    const outputPath = null;
    fs.ensureDirSync(sourcePath);
    const expectedError = resourcesHelpers.formatError(resource, errors.OUTPUT_PATH_NOT_GIVEN);

    // when
    expect(() => taArticleHelpers.processTranslationAcademy(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

});
