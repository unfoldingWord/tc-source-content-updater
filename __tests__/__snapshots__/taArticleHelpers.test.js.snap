// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for taArticleHelpers Test taArticleHelpers.processTranslationAcademy() for en 1`] = `
Array [
  "authority-level1.md",
  "authority-level2.md",
  "authority-level3.md",
  "authority-process.md",
  "self-assessment.md",
  "self-check.md",
]
`;

exports[`Tests for taArticleHelpers Test taArticleHelpers.processTranslationAcademy() for en 2`] = `
Array [
  "biblicalimageryta.md",
  "figs-grammar.md",
  "figs-metaphor.md",
  "qualifications.md",
  "translate-whatis.md",
]
`;

exports[`Tests for taArticleHelpers Test taArticleHelpers.processTranslationAcademy() for en 3`] = `
"# What is Translation? #

### Definition

Translation is a process performed between different languages that requires a person (the translator) to understand the meaning that a writer or speaker intended to communicate to an original audience in the source language, and then to express that same meaning to a different audience in the target language.

This is how translation is supposed to work most of the time, but sometimes certain translations have other goals, such as to reproduce the form of a source language, as we will see below.

There are basically two kinds of translations: literal and dynamic (or meaning-based).

* Literal translations focus on representing words in the source language with words in the target language that have similar basic meanings. They also use phrases that have similar structures to the phrases in the source language. This kind of translation allows the reader to see the structure of the source text, but it can make it difficult or impossible for the reader to understand the meaning of the source text.
* Dynamic, meaning-based translations focus on representing the meaning of the source language sentence in its context, and will use whatever words and phrase structures are most appropriate to convey that meaning in the target language. The goal of this kind of translation is to make it easy for the reader to understand the meaning of the source text. This is the kind of translation recommended in this Translation Manual for Other Language (OL) translations.

The ULB is designed to be a literal translation, so that the OL translator can see the forms of the original biblical languages. The UDB is designed to be a dynamic translation, so that the OL translator can understand the meaning of these forms in the Bible. When translating these resources, please translate the ULB in a literal way and translate the UDB in a dynamic way. For more information about these resources, see the Gateway Manual."
`;
